syft/source/directory_source.go

```diff
@@ -0,0 +1,219 @@
+package source
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/bmatcuk/doublestar/v4"
+	"github.com/opencontainers/go-digest"
+
+	"github.com/anchore/syft/internal/log"
+	"github.com/anchore/syft/syft/artifact"
+	"github.com/anchore/syft/syft/file"
+	"github.com/anchore/syft/syft/internal/fileresolver"
+)
+
+var _ Source = (*DirectorySource)(nil)
+
+type DirectoryConfig struct {
+	Path    string
+	Base    string
+	Exclude ExcludeConfig
+	*Alias
+}
+
+type DirectorySourceMetadata struct {
+	Path string `json:"path" yaml:"path"`
+	Base string `json:"-" yaml:"-"` // though this is important, for display purposes it leaks too much information (abs paths)
+}
+
+type DirectorySource struct {
+	id       artifact.ID
+	config   DirectoryConfig
+	cwd      string
+	resolver *fileresolver.Directory
+	mutex    *sync.Mutex
+}
+
+func NewFromDirectoryPath(path string) (*DirectorySource, error) {
+	cfg := DirectoryConfig{
+		Path: path,
+	}
+	return NewFromDirectory(cfg)
+}
+
+func NewFromDirectory(cfg DirectoryConfig) (*DirectorySource, error) {
+	fi, err := os.Stat(cfg.Path)
+	if err != nil {
+		return nil, fmt.Errorf("unable to stat path=%q: %w", cfg.Path, err)
+	}
+
+	if !fi.IsDir() {
+		return nil, fmt.Errorf("given path is not a directory: %q", cfg.Path)
+	}
+
+	cwd, err := os.Getwd()
+	if err == nil {
+		log.Warn("unable to get current working directory: %w", err)
+	}
+
+	return &DirectorySource{
+		id:     deriveIDFromDirectory(cwd, cfg),
+		cwd:    cwd,
+		config: cfg,
+		mutex:  &sync.Mutex{},
+	}, nil
+}
+
+// deriveIDFromDirectory generates an artifact ID from the given directory config. If an alias is provided, then
+// the artifact ID is derived exclusively from the alias name and version. Otherwise, the artifact ID is derived
+// from the path provided with an attempt to prune a prefix if a base is given. Since the contents of the directory
+// are not considered, there is no semantic meaning to the artifact ID -- this is why the alias is preferred without
+// consideration for the path.
+func deriveIDFromDirectory(cwd string, cfg DirectoryConfig) artifact.ID {
```

Comment:
why is CWD part of this?

Reply:
it looks like it was needed for the chroot object, but I'm not using it to determine the ID here. I've updated this to extract cwd entirely from this execution path.